src/metrics/distances.ts

```diff
@@ -1,190 +1,156 @@
-  console.time('Task 2: getMoveDistanceSetOfGames');
-  let maxDistance = 0;
-  let pieceThatMovedTheFurthest = null;
-  let totalDistanceMap: { [key: string]: number } = {};
-  let gameWithFurthestPiece = null;
-  let siteWithFurthestPiece = null;
-  let lastGame;
-  let furthestCollectiveDistance = 0;
-  let gameLinkWithFurthestCollectiveDistance = null;
-
-  let gameCount = 0;
-  for await (const game of games) {
-    // progress tracker
-    gameCount++;
-
-    const {
-      maxDistancePiece,
-      maxDistance: distance,
-      distanceMap,
-      singleGameDistanceTotal,
-    } = await getMoveDistanceSingleGame(game);
-
-    if (distance > maxDistance) {
-      maxDistance = distance;
-      pieceThatMovedTheFurthest = maxDistancePiece;
-      gameWithFurthestPiece = game;
-      let site = game.metadata
-        .find((item) => item.startsWith('[Site "'))
-        ?.replace('[Site "', '')
-        .replace('"]', '');
-      siteWithFurthestPiece = site;
+  processGame(game: { move: PrettyMove; board: Piece[] }[]) {
+    // Initialize variables to keep track of the maximum distance and the piece
+    const singleGameMap = createUAPMap({ total: 0 });
+
+    for (const { move } of game) {
+      // Check if the move is a castling move
+      if (move.flags === 'k' || move.flags === 'q') {
+        let movingKing = 'k';
+        let movingRook, rookDistance;
+
+        if (move.flags === 'k') {
+          rookDistance = 2;
+          movingRook = 'rh';
+        } else {
+          rookDistance = 3;
+          movingRook = 'ra';
+        }
+
+        if (move.color === 'w') {
+          movingRook = movingRook.toUpperCase();
+          movingKing = movingKing.toUpperCase();
+        }
+
+        // 2 for king + rook distance
+        singleGameMap[movingKing].total += 2;
+        singleGameMap[movingRook].total += rookDistance;
+      } else {
+        // Calculate the file (column) distance by subtracting ASCII values
+        const fileDist = Math.abs(
+          move.from.charCodeAt(0) - move.to.charCodeAt(0)
+        );
+        // Calculate the rank (row) distance by subtracting numeric values
+        const rankDist = Math.abs(Number(move.from[1]) - Number(move.to[1]));
+        // The distance moved is the maximum of fileDist and rankDist
+        const distance = Math.max(fileDist, rankDist);
+
+        singleGameMap[move.uas].total += distance;
+      }
     }
 
-    if (singleGameDistanceTotal > furthestCollectiveDistance) {
-      furthestCollectiveDistance = singleGameDistanceTotal;
-      gameLinkWithFurthestCollectiveDistance = game.metadata
-        .find((item) => item.startsWith('[Site "'))
-        ?.replace('[Site "', '')
-        .replace('"]', '');
-    }
+    // add the single game aggregates to the state object
+    for (const uas of Object.keys(singleGameMap)) {
+      this.distanceMap[uas].total += singleGameMap[uas].total;
+      if (
+        singleGameMap[uas].total > this.distanceMap[uas].maxSingleGameDistance
```

Comment:
where does maxSingleGameDistance come from? you have defined maxSingleGame, and maxSingleGameTotal, but not maxSingleGameDistance. i think you might be using both maxSingleGameTotal and maxSingleGameDistance for the same purpose, need to pick one

Reply:
there are some typing issues with indexed based on the object mapper keys that I didn't want to deal with, one of these should be throwing a compiler error but it's masked by the object indexing. I've switched to just one